### PR TITLE
adding a simple validator to the attrs class 

### PR DESCRIPTION
**Pull request by djarecka**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Summary
adding a type validator for attr classes  (not finished) - I had to write a wrapper around attr validators, since we are using `ty.Union` and `ty.List/Dict` and the validators can't check the types automatically. 
Decided to install `typing-inspect` to get the types from `ty.Union` etc., for py3.8 `typing` has its own method - `get_args`, but not for py3.7.
I'm not checking all combinations for now, i.e. I do not use recursion to check all depths of the type.

Checking values - for now only `allowed_values` ia added to the allowed metadata keys, but in the future it can be used for other validations.

## Checklist

- [ ] All tests passing
- [ ] I have added tests to cover my changes
- [ ] I have updated documentation (if necessary)
- [ ] My code follows the code style of this project
(we are using `black`: you can `pip install pre-commit`,
run `pre-commit install` in the `pydra` directory
and `black` will be run automatically with each commit)

## Acknowledgment
- [x] I acknowledge that this contribution will be available under the Apache 2 license.

### CODE REVIEW

File: pydra/engine/helpers.py [hunk 1]
```diff
@@ -13,9 +13,13 @@
 import re
 from time import strftime
 from traceback import format_exception
+import typing as ty
+import typing_inspect as tyi
```
Comment (satra) [inline pydra/engine/helpers.py:17]: can we drop this and code up these inspections ourselves? or use attrs?

Comment (satra) [reply pydra/engine/helpers.py:17]: we do this in other parts of the code where we gather type information.

Comment (djarecka) [reply pydra/engine/helpers.py:17]: if you have an example, happy to see

**Comment by djarecka** [comment]
@satra - is this `on/off` flag should be on the tasks level, or you're thinking about `pydra` settings?

**Comment by satra** [comment]
for on/off flag, it should be global (but with the potential to turn it on for a specific task if need be). but to start with at least global. see the last example here: https://www.attrs.org/en/stable/init.html#callables

for typing. attrs itself has direct support for this: https://www.attrs.org/en/stable/types.html and more specifically: https://www.attrs.org/en/stable/examples.html?highlight=validation#types

**Comment by djarecka** [comment]
sure I can use typing in the classes, but their validators do not work well with types like `ty.List[int]`, try:
```
@attr.s
class OtherClass:
    a_number = attr.ib(validator=attr.validators.instance_of(int))
    list_of_numbers = attr.ib(validator=attr.validators.instance_of(ty.List[int]))
bb = OtherClass(a_number=3, list_of_numbers=[1, 2])
```

**Comment by satra** [comment]
i'm simply saying that the information about typing can be retrieved through attrs (instead of the typing_inspect package), i'm not saying that you don't need to create custom validation routes for them.

**Comment by djarecka** [comment]
if you know how to do it, could you please share it with me, because I did have a problem how to get `int` from `ty.List[int]`, or how to get `(str, int)` from `ty.Union[str, int]`.

**Comment by satra** [comment]
```
In [22]: import typing 
    ...: @attr.s(auto_attribs=True) 
    ...: class AutoC: 
    ...:     cls_var: typing.ClassVar[int] = 5  # this one is ignored 
    ...:     l: typing.List[int] = attr.Factory(list) 
    ...:     x: int = 1 
    ...:     foo: str = attr.ib( 
    ...:          default="every attrib needs a type if auto_attribs=True" 
    ...:     ) 
    ...:     bar: typing.Any = None 
    ...:     zzz: typing.Optional[typing.Union[typing.List[str], int]] = ["a"] 
    ...:                                                                                        

In [23]: attr.fields(AutoC).zzz.type.__dict__                                                   
Out[23]: 
{'_inst': True,
 '_special': False,
 '_name': None,
 '__origin__': typing.Union,
 '__args__': (typing.List[str], int, NoneType),
 '__parameters__': (),
 '__slots__': None,
 '__module__': 'typing'}

In [24]: attr.fields(AutoC).l.type.__dict__                                                     
Out[24]: 
{'_inst': False,
 '_special': False,
 '_name': 'List',
 '__origin__': list,
 '__args__': (int,),
 '__parameters__': (),
 '__slots__': None}

```

**Comment by djarecka** [comment]
ok, so i've removed `typing_inspect`, using `__args__`, `__origin__` etc. for python 3.7, but for python 3.8 we methods from `typing_inspect` are already inside `typing`

**Comment by djarecka** [comment]
don't follow, so you wouldn't use the `validator` field in `attr.ib`?

**Pull request merged**